### PR TITLE
i#3044 AArch64 SVE codec: change LDR/STR and PRF to use byte offsets

### PR DESCRIPTION
**Pull request by AssadHashmi**
For the current decode/encode functions of:
```
LDR <Zt>, [<Xn|SP>{, #<imm>, MUL VL}]
LDR <Pt>, [<Xn|SP>{, #<imm>, MUL VL}]
STR <Zt>, [<Xn|SP>{, #<imm>, MUL VL}]
STR <Pt>, [<Xn|SP>{, #<imm>, MUL VL}]
PRFB <prfop>, <Pg>, [<Xn|SP>{, #<imm>, MUL VL}]
PRFH <prfop>, <Pg>, [<Xn|SP>{, #<imm>, MUL VL}]
PRFW <prfop>, <Pg>, [<Xn|SP>{, #<imm>, MUL VL}]
PRFD <prfop>, <Pg>, [<Xn|SP>{, #<imm>, MUL VL}]
```
Vector indexing is used in the memory operand at the IR level. However
the IR must always refer to the address in terms of the base register
value plus a byte offset displacement. This patch changes the
decode/encode functions for these instructions to expect byte offsets
at the IR level, converting to vector length offsets within the codec.

Issues #3044, #5365

### CODE REVIEW

File: core/ir/aarch64/codec.c [hunk 1]
```diff
@@ -5193,9 +5203,20 @@ encode_opnd_svemem_gpr_simm6_vl(uint enc, int opcode, byte *pc, opnd_t opnd,
...
         return false;
+    if (!reg_is_gpr(opnd_get_base(opnd)))
+        return false;
+
+    /* As described in decode_opnd_svemem_gpr_simm6_vl(), disp is a multiple of
+     * vector length at the IR level, transformed to a vector index in the
+     * encoding.
+     */
+    int vl_bytes = dr_get_sve_vl() / 8;
+    ASSERT((opnd_get_disp(opnd) % vl_bytes) == 0);
```
Comment (derekbruening) [inline core/ir/aarch64/codec.c:5214]: Best to just return false and fail the encoding like all the other un-encodable cases checked in this routine, rather than assert, if it's not an even multiple.

File: core/ir/aarch64/codec.c [hunk 2]
```diff
@@ -5321,16 +5355,29 @@ encode_opnd_svemem_gpr_simm9_vl(uint enc, int opcode, byte *pc, opnd_t opnd,
...
-    if (!encode_reg(&rn, &is_x, opnd_get_base(opnd), true) || !is_x)
-        return false;
+    /* As described in decode_opnd_svemem_gpr_simm9_vl(), disp is a multiple of
+     * vector or predicate length at the IR level, transformed to a vector or
+     * predicate index in the encoding.
+     */
+    int vl_bytes = dr_get_sve_vl() / 8;
+    int pl_bytes = vl_bytes / 8;
+    if (is_vector)
+        ASSERT((opnd_get_disp(opnd) % vl_bytes) == 0);
```
Comment (derekbruening) [inline core/ir/aarch64/codec.c:5372]: Like above I would have both of these just return false instead of asserting.

File: suite/tests/api/ir_aarch64_sve.c [hunk 3]
```diff
@@ -8303,23 +8303,23 @@ TEST_INSTR(compact_sve)
 TEST_INSTR(str)
 {
     /* Testing STR <Zt>, [<Xn|SP>{, #<simm>, MUL VL}] */
-    int simm_0[6] = { 0, 255, -256, 127, -128, -1 };
+    int simm_0[6] = { 0, 32, -32, 8160, -4096, -8192 };
```
Comment (derekbruening) [inline suite/tests/api/ir_aarch64_sve.c:8306]: Before it seemed to be checking boundary conditions: but the new numbers don't look like the same conditions?  Or at least they aren't a simple fixed scale-up.  Ditto below.  How did you come up with the new numbers?

**Review by derekbruening — APPROVED** [review]

File: core/ir/aarch64/codec.c (hunk 1 above)
Comment (AssadHashmi) [reply core/ir/aarch64/codec.c:5214]: Done.

File: core/ir/aarch64/codec.c (hunk 2 above)
Comment (AssadHashmi) [reply core/ir/aarch64/codec.c:5372]: Done.

File: suite/tests/api/ir_aarch64_sve.c (hunk 3 above)
Comment (AssadHashmi) [reply suite/tests/api/ir_aarch64_sve.c:8306]: For test vector length of 256 bits the `simm_0` array does include boundary values:
255 x 32 bytes -> 8160
-256 x 32 bytes -> -8192

Updated two test values for `STR <Pt>, [<Xn|SP>{, #<simm>, MUL VL}]` and `LDR <Zt>, [<Xn|SP>{, #<simm>, MUL VL}]` which weren't boundary values.

Comment (derekbruening) [reply suite/tests/api/ir_aarch64_sve.c:8306]: OK, I was just expecting the same scale factor in the diff: so if -256 became -32, -128 should become -16, but it became -4096 instead.

**Review by derekbruening — APPROVED** [review]
